CMSgov/design-system
# [WNMGDS-422] Create release scripts

### Summary
- Updates our package.json to include a new npm script (`yarn release`). This script calls an updated `prepublish.sh` and then `lerna publish`. 
- Updates `prepublish.sh` to do the following:
   - Pulls the latest from GH
   - Cleans `node_modules` and runs a fresh install
   - Builds files
   - Runs unit tests, e2e tests, linter (`backstopjs` is not run, as we are still working that out)
   - Runs `lerna version`, which will bump the package versions, author and tag a release commit, and push this commit up to a new release branch.
   - Create a release zip to include in the release notes
- Update our release notes

### Notes
- These changes are currently internal to our team, and I didn't generalize them for child DS's yet. We use `lerna` and have some additional steps to our release process, so I thought it was better to add a recommended publishing process for child DS's later.
- The new release script doesn't contain 2 steps: `backstopJS` and the release notes. While there is a way to create a release via the command line, I thought it was best to keep this first iteration simple, and mostly rely on `lerna` for automating this process.

### How to test
In order to test `prepublish.sh`, first you need to:
   1. Remove the `"command"` entry in `lerna.json`. The `allowBranch` property prevents `lerna` commands to be run on any non `master` branches.
   2. Comment out line 10-11 in `prepublish.sh`. 

After all these steps, you should be able to run `./prepublish.sh`. After all the tests pass and `lerna version` is run, you should be able to run `git log` to see the new release commit in your local git history and you should be on a new branch with the following naming convention `release-{version number}`. If you run `git tag`, you should also be able to find the newly created tag with the same name as the branch.

NOTE: This script pushes a release branch and tag to Github, so after testing you will want to undo these changes. 
1. Delete the newly created release branch from the repo [here](https://github.com/CMSgov/design-system/branches)
2. Delete the new tag from origin `git push --delete origin [tag name]`
3. Delete the new tag locally `git tag -d [tag name]`
